### PR TITLE
CPLAT 6300 React forwardRef

### PR DESCRIPTION
**Pull request by sydneyjodon-wk**
Hold: This PR is reliant upon the interop introduced in [react-dart#193](https://github.com/cleandart/react-dart/pull/193) within react-dart. Until it is merged, the CI will fail
## Motivation
  
Expose React 16 `forwardRef` in `over_react`. `forwardRef` is a technique for automatically passing a ref through a component to one of its children.
## Changes
  
Create wrapper function for `forwardRef`.
## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 
@kealjones-wk @aaronlademann-wf @greglittlefield-wf @joebingham-wk 
### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
      - [ ] CI passes
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/component/forward_ref.dart [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+//// limitations under the License.
+
+import 'package:react/react_client.dart' as react;
+import 'package:over_react/component_base.dart';
+
+UiFactory<TProps> Function(UiFactory<TProps>) forwardRef<TProps extends UiProps>(
+    Function(TProps props, dynamic ref) wrapperFunction) {
+
+  UiFactory<TProps> wrapWithForwardRef(UiFactory<TProps> factory) {
+    wrapProps(Map props, ref) {
```
Comment (greglittlefield-wf) [inline lib/src/component/forward_ref.dart]: This should have a return type.

File: test/over_react/component/forward_ref_test.dart [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    if (hasProperty(refValue.current, 'id')) {
+      idValue = refValue.current.id;
+    } else {
+      idValue = refValue.current.props['id'];
+    }
+
+    expect(idValue, equals('test'), reason: 'child component should have access to parent props');
+    verifyRefValue(refValue.current);
+  });
+
```
Comment (joebingham-wk) [inline test/over_react/component/forward_ref_test.dart]: #nit extra line

File: lib/src/component/ref_util.dart [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+///         ..className = 'special-class'
+///       )(
+///         props.children
+///       );
+///     })(Dom.div);
+///
+/// ___ OR ___
+///
+///     UiFactory<FooProps> FooForwarded = forwardRef<FooProps>((props, ref) {
+///       return Dom.form()(
```
Comment (greglittlefield-wf) [inline lib/src/component/ref_util.dart]: Shouldn't this example render a `Foo`?

Comment (sydneyjodon-wk) [reply lib/src/component/ref_util.dart]: Good point, you're right. I'll fix that!

File: pubspec.yaml [hunk 4]
```diff
@@ -42,4 +42,4 @@ dependency_overrides:
   react:
     git:
       url: https://github.com/cleandart/react-dart.git
-      ref: 5.1.0-wip
+      ref: CPLAT-4337-react-createRef-forwardRef
```
Comment (greglittlefield-wf) [inline pubspec.yaml]: This needs to be reverted now that the react-dart PR has merged

**Pull request merged**